File: content/examples/other/leetcode/problems/house-robber/example-1/solution.py
def rob_top_down(nums):
    memo = {}

    def dfs(i):
        if i >= len(nums):
            return 0
        if i in memo:
            return memo[i]
        best = max(dfs(i + 1), nums[i] + dfs(i + 2))
        memo[i] = best
        return best

    return dfs(0)


if __name__ == "__main__":
    print(rob_top_down([1, 2, 3, 1]))

File: content/examples/other/leetcode/problems/house-robber/example-1/solution.js
function robTopDown(nums) {
  const memo = new Map();
  function dfs(i) {
    if (i >= nums.length) return 0;
    if (memo.has(i)) return memo.get(i);
    const best = Math.max(dfs(i + 1), nums[i] + dfs(i + 2));
    memo.set(i, best);
    return best;
  }
  return dfs(0);
}

if (require.main === module) {
  console.log(robTopDown([1, 2, 3, 1]));
}

module.exports = { robTopDown };
